python/tvm/relay/ir_pass.py

```diff
@@ -1,16 +1,287 @@
 # pylint: disable=no-else-return
 # pylint: disable=unidiomatic-typecheck
-"""The set of passes for Relay.
+"""
+This file contains:
+1. The set of passes for Relay, which exposes an interface for configuring the
+   passes and scripting them in Python.
 
-Exposes an interface for configuring the passes and
-scripting them in Python.
+2. The pass manager for Relay which exposes different granularity of interfaces
+   for users to implement and use passes more conveniently.
 """
+import types
+
 from . import _ir_pass
 from . import _make
 from .expr import Expr
 from .ty import Type
+from .base import RelayNode, register_relay_node
 from .module import Module
 
+
+@register_relay_node
+class PassContext(RelayNode):
+    """The basis where a Relay optimization/analysis runs on.
+    Each pass context contains a number of auxiliary information that is used
+    to help an optimization pass. Such information includes the error reporter
+    to record the errors of during the performing the optimization, etc.
+    """
+
+    def __init__(self):
+        self.__init_handle_by_constructor__(_ir_pass.PassContext)
+
+
+@register_relay_node
+class Pass(RelayNode):
+    """The base class of all passes. This class is designed as a pure virtual
+    class that will be implemented by the subclasses.
+
+    Parameters
+    ----------
+    name : str
+        The pass name.
+
+    opt_level : int
+        The optimization level of this pass.
+    """
+
+    def set_pass_context(self, pass_ctx):
+        """Setup the pass context for analysis and optimizations. This context
+        could be shared by different passes for sequential passes.
+
+        Parameters
+        ----------
+        pass_ctx : PassContext
+            The context that is used to help perform a certain pass or a series
+            of passes.
+
+        Returns
+        -------
+        pass : Pass
+            The updated pass.
+        """
+        if not isinstance(pass_ctx, PassContext):
+            raise TypeError("pass_ctx is expected to be the PassContext type")
+        return _ir_pass.SetContext(self, pass_ctx)
+
+    def __call__(self, mod):
+        """Execute the pass. It is an abstract function that will be
+        implemented by subclasses.
+
+        Parameters
+        ----------
+        mod : tvm.relay.Module
+            The module that a certain optimization is performed on.
+
+        Returns
+        -------
+        mod : tvm.relay.Module
+            The updated module after applying this pass.
+        """
+        raise NotImplementedError("Pure virtual function is not implemented.")
+
+
+@register_relay_node
+class ModulePass(Pass):
+    """A pass that works on tvm.relay.Module.
+
+    Parameters
+    ----------
+    name : str
+        The pass name.
+
+    opt_level : int
+        The optimization level of this pass.
+
+    pass_func : Callable[PassContext: tvm.relay.Module -> tvm.relay.Module]
```

Comment:
```suggestion
    pass_func : Callable[(tvm.relay.Module, PassContext) -> tvm.relay.Module]
```

Is this what you mean?